File: src/main/java/org/cubexmc/railway/util/OwnershipUtil.java
package org.cubexmc.railway.util;

import java.util.Objects;
import java.util.Set;
import java.util.UUID;

import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.cubexmc.railway.model.Line;
import org.cubexmc.railway.model.Stop;

/**
 * Utility methods for handling ownership and permission checks on lines and stops.
 */
public final class OwnershipUtil {

    public static final String PERMISSION_ADMIN = "railway.admin";
    public static final String PERMISSION_LINE_CREATE = "railway.line.create";
    public static final String PERMISSION_STOP_CREATE = "railway.stop.create";

    private OwnershipUtil() {
    }

    public static boolean hasAdminBypass(CommandSender sender) {
        return !(sender instanceof Player) || sender.hasPermission(PERMISSION_ADMIN);
    }

    public static boolean canCreateLine(CommandSender sender) {
        return hasAdminBypass(sender) || sender.hasPermission(PERMISSION_LINE_CREATE);
    }

    public static boolean canCreateStop(CommandSender sender) {
        return hasAdminBypass(sender) || sender.hasPermission(PERMISSION_STOP_CREATE);
    }

    public static boolean isServerOwned(Line line) {
        return line != null && line.getOwner() == null;
    }

    public static boolean isServerOwned(Stop stop) {
        return stop != null && stop.getOwner() == null;
    }

    public static boolean isLineAdmin(UUID playerId, Line line) {
        if (line == null || playerId == null) return false;
        Set<UUID> admins = line.getAdmins();
        return admins.contains(playerId);
    }

    public static boolean isStopAdmin(UUID playerId, Stop stop) {
        if (stop == null || playerId == null) return false;
        Set<UUID> admins = stop.getAdmins();
        return admins.contains(playerId);
    }

    public static boolean canManageLine(CommandSender sender, Line line) {
        if (hasAdminBypass(sender)) return true;
        if (!(sender instanceof Player player) || line == null) return false;
        if (isServerOwned(line)) return player.isOp();
        return isLineAdmin(player.getUniqueId(), line);
    }

    public static boolean canManageStop(CommandSender sender, Stop stop) {
        if (hasAdminBypass(sender)) return true;
        if (!(sender instanceof Player player) || stop == null) return false;
        if (isServerOwned(stop)) return player.isOp();
        return isStopAdmin(player.getUniqueId(), stop);
    }

    public static boolean canModifyLineStops(CommandSender sender, Line line, Stop stop) {
        if (!canManageLine(sender, line) || stop == null) return false;
        if (hasAdminBypass(sender)) return true;
        if (!(sender instanceof Player player)) return false;
        UUID playerId = player.getUniqueId();
        if (isStopAdmin(playerId, stop)) return true;
        if (isServerOwned(stop)) return player.isOp();
        return stop.isLineAllowed(line.getId());
    }

    public static boolean canLinkStopToLine(CommandSender sender, Line line, Stop stop) {
        if (hasAdminBypass(sender)) return true;
        if (!(sender instanceof Player player) || line == null || stop == null) return false;
        UUID playerId = player.getUniqueId();
        if (isServerOwned(line) && !player.isOp()) return false;
        if (isServerOwned(stop)) return player.isOp();
        if (!isLineAdmin(playerId, line)) return false;
        if (isStopAdmin(playerId, stop)) return true;
        return stop.isLineAllowed(line.getId());
    }

    public static boolean canLinkLineWithoutPlayer(Line line, Stop stop) {
        if (line == null || stop == null) return false;
        if (stop.isLineAllowed(line.getId())) return true;
        return Objects.equals(stop.getOwner(), line.getOwner());
    }
}

File: src/main/java/org/cubexmc/metro/util/OwnershipUtil.kt
package org.cubexmc.metro.util

import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.model.Portal
import org.cubexmc.metro.model.Stop
import java.util.Objects
import java.util.UUID

/**
 * 所有权与权限检查工具类，封装线路和站点的管理逻辑。
 */
object OwnershipUtil {
    const val PERMISSION_ADMIN: String = "railway.admin"
    const val PERMISSION_LINE_CREATE: String = "railway.line.create"
    const val PERMISSION_STOP_CREATE: String = "railway.stop.create"
    const val PERMISSION_PORTAL_CREATE: String = "railway.portal.create"

    @JvmStatic
    fun hasAdminBypass(sender: CommandSender): Boolean = sender !is Player || sender.hasPermission(PERMISSION_ADMIN)

    @JvmStatic
    fun canCreateLine(sender: CommandSender): Boolean = hasAdminBypass(sender) || sender.hasPermission(PERMISSION_LINE_CREATE)

    @JvmStatic
    fun canCreateStop(sender: CommandSender): Boolean = hasAdminBypass(sender) || sender.hasPermission(PERMISSION_STOP_CREATE)

    @JvmStatic
    fun canCreatePortal(sender: CommandSender): Boolean = hasAdminBypass(sender) || sender.hasPermission(PERMISSION_PORTAL_CREATE)

    @JvmStatic
    fun isServerOwned(line: Line?): Boolean = line != null && line.owner == null

    @JvmStatic
    fun isServerOwned(stop: Stop?): Boolean = stop != null && stop.owner == null

    @JvmStatic
    fun isServerOwned(portal: Portal?): Boolean = portal != null && portal.owner == null

    @JvmStatic
    fun isLineAdmin(playerId: UUID?, line: Line?): Boolean = line != null && playerId != null && line.admins.contains(playerId)

    @JvmStatic
    fun isStopAdmin(playerId: UUID?, stop: Stop?): Boolean = stop != null && playerId != null && stop.admins.contains(playerId)

    @JvmStatic
    fun isPortalAdmin(playerId: UUID?, portal: Portal?): Boolean = portal != null && playerId != null && portal.admins.contains(playerId)

    @JvmStatic
    fun canManageLine(sender: CommandSender, line: Line?): Boolean {
        if (hasAdminBypass(sender)) {
            return true
        }
        if (sender !is Player || line == null) {
            return false
        }
        if (isServerOwned(line)) {
            return sender.isOp
        }
        return isLineAdmin(sender.uniqueId, line)
    }

    @JvmStatic
    fun canManageStop(sender: CommandSender, stop: Stop?): Boolean {
        if (hasAdminBypass(sender)) {
            return true
        }
        if (sender !is Player || stop == null) {
            return false
        }
        if (isServerOwned(stop)) {
            return sender.isOp
        }
        return isStopAdmin(sender.uniqueId, stop)
    }

    @JvmStatic
    fun canManagePortal(sender: CommandSender, portal: Portal?): Boolean {
        if (hasAdminBypass(sender)) {
            return true
        }
        if (sender !is Player || portal == null) {
            return false
        }
        if (isServerOwned(portal)) {
            return sender.isOp
        }
        return isPortalAdmin(sender.uniqueId, portal)
    }

    @JvmStatic
    fun canModifyLineStops(sender: CommandSender, line: Line?, stop: Stop?): Boolean {
        if (!canManageLine(sender, line) || stop == null) {
            return false
        }
        if (hasAdminBypass(sender)) {
            return true
        }
        val currentLine = line ?: return false
        if (sender !is Player) {
            return false
        }
        val playerId = sender.uniqueId
        if (isStopAdmin(playerId, stop)) {
            return true
        }
        if (isServerOwned(stop)) {
            return sender.isOp
        }
        return stop.isLineAllowed(currentLine.id)
    }

    @JvmStatic
    fun canLinkStopToLine(sender: CommandSender, line: Line?, stop: Stop?): Boolean {
        if (hasAdminBypass(sender)) {
            return true
        }
        if (sender !is Player || line == null || stop == null) {
            return false
        }
        val playerId = sender.uniqueId

        if (isServerOwned(line) && !sender.isOp) {
            return false
        }
        if (isServerOwned(stop)) {
            return sender.isOp
        }
        if (!isLineAdmin(playerId, line)) {
            return false
        }
        if (isStopAdmin(playerId, stop)) {
            return true
        }
        return stop.isLineAllowed(line.id)
    }

    @JvmStatic
    fun canLinkLineWithoutPlayer(line: Line?, stop: Stop?): Boolean {
        if (line == null || stop == null) {
            return false
        }
        if (stop.isLineAllowed(line.id)) {
            return true
        }
        return Objects.equals(stop.owner, line.owner)
    }
}
